compiler/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -166,6 +220,293 @@ impl Context {
+        return;
+    }
+
+    // What's the last instruction that's part of the group? (5 in the example above)
+    let last_instruction_index = index + 2 * (element_size - 1);
+    // How many unused instructions are in this group?
+    let mut unused_count = 1;
+    loop {
+        *next_out_of_bounds_index = possible_index_out_of_bounds_indexes.pop();
+        if let Some(out_of_bounds_index) = *next_out_of_bounds_index {
+            if out_of_bounds_index <= last_instruction_index {
+                unused_count += 1;
+                if unused_count == element_size {
+                    // We are in case b): we need to insert just one constrain.
+                    // Since we popped all of the group indexes, and given that we
+                    // are analyzing the first instruction in the group, we can
+                    // set `next_out_of_bounds_index` to the current index:
+                    // then a check will be inserted, and no other check will be
+                    // inserted for the rest of the group.
+                    *next_out_of_bounds_index = Some(index);
+                    break;
+                } else {
+                    continue;
+                }
+            }
+        }
+
+        // We are in case c): some of the instructions are unused.
+        // We don't need to insert any checks, and given that we already popped
+        // all of the indexes in the group, there's nothing else to do here.
+        break;
+    }
+}
+
+// Given `lhs` and `rhs` values, if there's a side effects condition this will
+// return (`lhs * condition`, `rhs * condition`), otherwise just (`lhs`, `rhs`)
+fn apply_side_effects(
+    side_effects_condition: Option<ValueId>,
+    lhs: ValueId,
+    rhs: ValueId,
+    function: &mut Function,
+    block_id: BasicBlockId,
+    call_stack: Vector<Location>,
+) -> (ValueId, ValueId) {
+    // See if there's an active "enable side effects" condition
+    let Some(condition) = side_effects_condition else {
+        return (lhs, rhs);
+    };
+
+    // Condition needs to be cast to argument type in order to multiply them together.
+    // In our case, lhs is always a boolean.
+    let casted_condition = function
+        .dfg
+        .insert_instruction_and_results(
+            Instruction::Cast(condition, Type::bool()),
+            block_id,
+            None,
+            call_stack.clone(),
+        )
+        .first();
+    let lhs = function
+        .dfg
+        .insert_instruction_and_results(
+            Instruction::binary(BinaryOp::Mul, lhs, casted_condition),
+            block_id,
+            None,
+            call_stack.clone(),
+        )
+        .first();
+    let rhs = function
+        .dfg
+        .insert_instruction_and_results(
+            Instruction::binary(BinaryOp::Mul, rhs, casted_condition),
+            block_id,
+            None,
+            call_stack,
+        )
+        .first();
```

Comment:
Here's the formatting issue I mentioned in the call

Reply:
Done! Let me know if it looks good now.